elytron/WFCORE-4994-case-principal-transformer.adoc

```diff
@@ -0,0 +1,119 @@
+:author:            Sonia Zaldana Calles
+:email:             szaldana@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Elytron currently has six principal transformers enabling aggregating principal transofmers, chaining them, using regular expressions, etc.
+However, the only way to adjust the case of a principal is by using a custom principal transformer as described
+https://developer.jboss.org/people/fjuma/blog/2019/02/25/implementing-a-custom-elytron-principal-transformer[here].
+
+It would be good to have a principal transformer to adjust a username to upper or lower case without having to go through
+the complicated process of creating a custom one.
+
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFCORE-4994[WFCORE-4994]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-13570[WFLY-13570]
+* https://issues.redhat.com/browse/ELY-1987[ELY-1987]
+* https://issues.jboss.org/browse/EAP7-1203[EAP7-1203]
+
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+TBD
+
+=== Affected Projects or Components
+
+* Wildfly, Wildfly-Core and Elytron
+
+=== Other Interested Projects
+
+N/A
+
+== Requirements
+
+=== Hard Requirements
+
+* A new ``case-principal-transformer`` resource needs to be added to the Elytron subsystem, which should be able
+to be configured in the ``mappers`` configuration. In particular, this new resource will have the following
+optional attribute:
+
+** ``upper-case`` - a boolean value outlining whether the principal should be transformed to all uppercase.
+This value will be set to ``true`` by default.
+
+* It must be possible to create a case-principal transformer which adjusts the principal to uppercase
+by default as follows:
+
+[source]
+----
+/subsystem=elytron/case-principal-transformer=my-case-transformer:add()
+----
+
+* It must be possible to adjust to lowercase by setting the attribute ``upper-case`` to false:
+[source]
+----
+/subsystem=elytron/case-principal-transformer=my-case-transformer:add(upper-case=false)
+----
+
+* It must be possible to use our ``case-principal-transformer`` in any place where we can use our other provided
+principal transformers. For example:
+
+[source]
+----
+/subsystem=elytron/security-domain=ManagementDomain:write-attribute(name=pre-realm-principal-transformer,
+value=my-case-transformer)
+----
+
+
+=== Nice-to-Have Requirements
+
+N/A
+
+=== Non-Requirements
+
+N/A
+
+
+== Test Plan
+
+* WildFly Elytron test suite: tests for case principal transformer functionality.
+* Wildfly Core test suite: testing of case principal transformer functionality when it is
```

Comment:
It would be good to mention that Elytron subsystem parsing and transformer tests will be added. Tests should also be added in the WildFly testsuite to ensure that the correct principal is used when a `case-principal-transformer` has been configured.

Reply:
fixed!